src/pcap/bosh.go

```diff
@@ -1,55 +1,414 @@
 package pcap
 
 import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
 	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
 
+	"github.com/golang-jwt/jwt/v4"
+	log "github.com/sirupsen/logrus"
 	"go.uber.org/zap"
 )
 
-type BoshHandler struct {
-	// TODO: add specifics for BOSH director configuration
-	Config ManualEndpoints
+type BoshInfo struct {
+	Name            string `json:"name"`
+	UUID            string `json:"uuid"`
+	Version         string `json:"version"`
+	Cpi             string `json:"cpi"`
+	StemcellOS      string `json:"stemcell_os"`
+	StemcellVersion string `json:"stemcell_version"`
+
+	UserAuthentication struct {
+		Type    string `json:"type"`
+		Options struct {
+			URL  string   `json:"url"`
+			URLs []string `json:"urls"`
+		} `json:"options"`
+	} `json:"user_authentication"`
+}
+
+type BoshInstance struct {
+	AgentID     string    `json:"agent_id"`
+	Cid         string    `json:"cid"`
+	Job         string    `json:"job"`
+	Index       int       `json:"index"`
+	ID          string    `json:"id"`
+	Az          string    `json:"az"`
+	Ips         []string  `json:"ips"`
+	VMCreatedAt time.Time `json:"vm_created_at"`
+	ExpectsVM   bool      `json:"expects_vm"`
+}
+
+type BoshResolverConfig struct {
+	RawDirectorURL   string     `yaml:"director_url" validate:"required,url"`
+	EnvironmentAlias string     `yaml:"alias" validate:"required"`
+	AgentPort        int        `yaml:"agent_port" validate:"required,gt=0,lte=65535"`
+	TokenScope       string     `yaml:"token_scope" validate:"required"`
+	MTLS             *MutualTLS `yaml:"mtls" validate:"omitempty"`
+}
+
+type BoshResolver struct {
+	client      *http.Client
+	uaaURLs     []string
+	config      BoshResolverConfig
+	directorURL *url.URL
+	logger      *zap.Logger
+	boshRootCAs *x509.CertPool
+}
+
+func NewBoshResolver(config BoshResolverConfig) (*BoshResolver, error) {
+	directorURL, err := url.Parse(config.RawDirectorURL)
+	if err != nil {
+		return nil, fmt.Errorf("cannot initialize BoshResolver for environment %s. %w", config.EnvironmentAlias, err)
+	}
+
+	var boshRootCAs *x509.CertPool
+	if config.MTLS != nil {
+		boshRootCAs, err = createCAPool(config.MTLS.CertificateAuthority)
+		if err != nil {
+			return nil, fmt.Errorf("could not create bosh CA pool: %w", err)
+		}
+	}
+
+	resolver := &BoshResolver{
+		logger:      zap.L().With(zap.String(LogKeyHandler, config.EnvironmentAlias)),
+		config:      config,
+		directorURL: directorURL,
+		boshRootCAs: boshRootCAs,
+	}
+
+	err = resolver.setup()
+	if err != nil {
+		return nil, err
+	}
+	return resolver, err
 }
 
-func (bosh *BoshHandler) name() string {
-	return "bosh"
+func (br *BoshResolver) Name() string {
+	return fmt.Sprintf("bosh/%s", br.config.EnvironmentAlias)
 }
 
-func (bosh *BoshHandler) canResolve(request *Capture) bool {
-	return request.GetBosh() != nil
+func (br *BoshResolver) CanResolve(request *EndpointRequest) bool {
+	if boshRequest := request.GetBosh(); boshRequest != nil {
```

Comment:
Is it possible that incoming _request_ is nil?

Reply:
I will add a check. But it will just return `false` in that case because no resolver can resolve an empty request.